controller/volume_rebuilding_controller.go

```diff
@@ -391,6 +456,63 @@ func (vbc *VolumeRebuildingController) isVolumeReplicasHealthy(numberOfReplicas
 	return healthyCount >= numberOfReplicas
 }
 
+func (vbc *VolumeRebuildingController) isVolumeReplicasReusableOrSchedulable(vol *longhorn.Volume, rs map[string]*longhorn.Replica) (bool, error) {
+	reusableFailedReplica, err := vbc.scheduler.CheckAndReuseFailedReplica(rs, vol, "")
+	if err != nil {
+		return false, err
+	}
+	if reusableFailedReplica != nil {
+		vbc.logger.Infof("Reusing failed replica %v for volume %v", reusableFailedReplica.Name, vol.Name)
+		return true, nil
+	}
+
+	hasReplicaDiskCandidates, err := vbc.hasReplicaDiskCandidates(vol, rs)
+	if err != nil {
+		return false, err
+	}
+	return hasReplicaDiskCandidates, nil
+}
+
+func (vbc *VolumeRebuildingController) hasReplicaDiskCandidates(vol *longhorn.Volume, rs map[string]*longhorn.Replica) (bool, error) {
+	engine, err := vbc.getVolumeEngine(vol)
+	if err != nil {
+		return false, err
+	}
+	if engine == nil {
+		vbc.logger.Warnf("Volume %v engine not found, skip offline rebuilding", vol)
+		return false, nil
+	}
+	newReplica := &longhorn.Replica{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            types.GenerateReplicaNameForVolume(vol.Name),
+			OwnerReferences: datastore.GetOwnerReferencesForVolume(vol),
+		},
+		Spec: longhorn.ReplicaSpec{
+			InstanceSpec: longhorn.InstanceSpec{
+				VolumeName:  vol.Name,
+				VolumeSize:  vol.Spec.Size,
+				Image:       vol.Status.CurrentImage,
+				DataEngine:  vol.Spec.DataEngine,
+				DesireState: longhorn.InstanceStateStopped,
+			},
+			EngineName:                       engine.Name,
+			Active:                           true,
+			BackingImage:                     vol.Spec.BackingImage,
+			HardNodeAffinity:                 "",
+			RevisionCounterDisabled:          vol.Spec.RevisionCounterDisabled,
+			UnmapMarkDiskChainRemovedEnabled: engine.Spec.UnmapMarkSnapChainRemovedEnabled,
+			SnapshotMaxCount:                 vol.Spec.SnapshotMaxCount,
+			SnapshotMaxSize:                  vol.Spec.SnapshotMaxSize,
+		},
+	}
```

Comment:
Seems redundance of `func (c *VolumeController) newReplica(v *longhorn.Volume, e *longhorn.Engine, hardNodeAffinity string) *longhorn.Replica {` Should we move that function to `util/volume.go` and use it for both rebuilding controller and volume controller

Reply:
Modified. Move the method `newReplica` to controller/utils.go as the method `newReplicaCR`.
Thanks.